casper/src/main/resources/PoS.rhox

```diff
@@ -86,24 +89,48 @@ new  PoS,
             } |
 
             contract PoS (@"pay", @amount, returnCh) = {
-              new vaultCh, transferCh, userCh, processCh, depositCh in {
+              new vaultCh, transferCh, userCh, processPayCh,
+                  depositCh,
+                  computeSum, totalStakeCh,
+                  computeDelta, rewardsDeltaCh,
+                  computeMergeMap, newRewardsCh in {
                 getUser!(*userCh) |
 
-                runMVar!(*stateCh, *processCh, *returnCh) |
+                runMVar!(*stateCh, *processPayCh, *returnCh) |
 
-                for(@(amounts, bonds), resultCh <- processCh;
+                for(@(rewards, bonds), payResultCh <- processPayCh;
                     @userPk <- userCh) {
                   deposit!(userPk, amount, *depositCh) |
+                  @ListOps!("fold", bonds.toList(), 0, *computeSum, *totalStakeCh) |
+                  contract computeSum(@(pk, (stake, _)), @acc, resultCh) = {
+                    resultCh!(acc + stake)
+                  } |
+
+                  for(@totalStake <- totalStakeCh) {
+                    @ListOps!("fold", bonds.toList(), {}, *computeDelta, *rewardsDeltaCh) |
+                    contract computeDelta(@(pk, (currentValidatorStake, _)), @acc, resultCh) = {
+                      resultCh!(acc.set(pk, (amount * currentValidatorStake) / totalStake))
+                    }
+                  } |
+
+                  for (@rewardsDelta <- rewardsDeltaCh) {
+                    @ListOps!("fold", rewardsDelta.toList(), rewards, *computeMergeMap, *newRewardsCh) |
+                    contract computeMergeMap(@(pk, rewardDelta), @newRewards, resultCh) = {
+                      resultCh!(newRewards.set(pk, newRewards.getOrElse(pk, 0) + rewardDelta))
+                    }
+                  } |
 
-                  for (@depositResult <- depositCh) {
-                    resultCh!((amounts, bonds), depositResult)
+                  for(@depositResult <- depositCh;
+                      @newRewards <- newRewardsCh) {
+                    payResultCh!((newRewards, bonds), depositResult)
                   }
                 }
               }
             } |
 
             contract PoS(@"slash", @blockHash, return) = {
```

Comment:
What happens to a validator's rewards when they are slashed? Should the rewards be set to zero?

Reply:
Right! I missed that part